rust/src/instruction/declaration.rs

```diff
@@ -0,0 +1,145 @@
+                        quil_rs::instruction::Vector::new(memory_type, memory_size),
+                        sharing,
+                    ),
+                ),
+            },
+        ))
+    }
+
+    #[getter]
+    fn memory_type(self_: PyRef<'_, Self>) -> PyResult<String> {
+        let instruction = self_.into_super();
+        let declaration = extract_instruction_as!(instruction, Declaration)?;
+        string_from_scalar_type(declaration.size.data_type)
+    }
+
+    #[setter]
+    fn set_memory_type(self_: PyRefMut<'_, Self>, memory_type: Bound<'_, PyAny>) -> PyResult<()> {
+        let mut instruction = self_.into_super();
+        let memory_type = conversion::py_to_scalar_type(&memory_type)?;
+        let declaration = extract_instruction_as_mut!(instruction, Declaration)?;
+        declaration.size.data_type = memory_type;
+        Ok(())
+    }
+
+    #[getter]
+    fn memory_size(self_: PyRef<'_, Self>) -> PyResult<u64> {
+        let instruction = self_.into_super();
+        let declaration = extract_instruction_as!(instruction, Declaration)?;
+        Ok(declaration.size.length)
+    }
+
+    #[setter]
+    fn set_memory_size(self_: PyRefMut<'_, Self>, memory_size: u64) -> PyResult<()> {
+        let mut instruction = self_.into_super();
+        let declaration = extract_instruction_as_mut!(instruction, Declaration)?;
+        declaration.size.length = memory_size;
+        Ok(())
+    }
+
+    #[getter]
+    fn shared_region(self_: PyRef<'_, Self>) -> PyResult<Option<String>> {
+        let instruction = self_.into_super();
+        let declaration = extract_instruction_as!(instruction, Declaration)?;
+        Ok(declaration.sharing.clone().map(|sharing| sharing.name))
+    }
+
+    #[setter]
+    fn set_shared_region(self_: PyRefMut<'_, Self>, shared_region: Option<String>) -> PyResult<()> {
+        let mut instruction = self_.into_super();
+        let declaration = extract_instruction_as_mut!(instruction, Declaration)?;
+        declaration.sharing = shared_region.map(|region_name| {
+            quil_rs::instruction::Sharing::new(
+                region_name,
+                declaration
+                    .sharing
+                    .take()
+                    .map(|sharing| sharing.offsets)
+                    .unwrap_or_default(),
+            )
+        });
+        Ok(())
+    }
+
+    #[getter]
+    fn offsets(self_: PyRef<'_, Self>) -> PyResult<Option<Vec<(u64, String)>>> {
+        let instruction = self_.into_super();
+        let declaration = extract_instruction_as!(instruction, Declaration)?;
+        Ok(match &declaration.sharing {
+            None => None,
+            Some(sharing) => optional_tuples_from_offsets(sharing.offsets.clone())?,
+        })
+    }
+
+    #[setter]
+    fn set_offsets(self_: PyRefMut<'_, Self>, offsets: Bound<'_, PyAny>) -> PyResult<()> {
+        let mut instruction = self_.into_super();
+        let offsets = conversion::py_to_offsets(&offsets)?;
```

Comment:
We have to manually run the conversion inside  #[setter] methods because using them with the `#[from_py_with="..."]` parameter attribute causes a compilation error. Opened pyo3/pyo3#3992 to track.